include/aws/mqtt/private/request-response/subscription_manager.h

```diff
@@ -0,0 +1,206 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/mqtt/mqtt.h>
+
+#include <aws/common/hash_table.h>
+
+struct aws_mqtt_protocol_adapter;
+struct aws_protocol_adapter_connection_event;
+struct aws_protocol_adapter_subscription_event;
+
+/*
+ * The kind of subscription event being emitted.
+ */
+enum aws_rr_subscription_event_type {
+
+    /*
+     * A subscribe succeeded
+     */
+    ARRSET_SUBSCRIPTION_SUBSCRIBE_SUCCESS,
+
+    /*
+     * A subscribe failed
+     */
+    ARRSET_SUBSCRIPTION_SUBSCRIBE_FAILURE,
+
+    /*
+     * A previously successful subscription has ended (generally due to a failure to resume a session)
+     */
+    ARRSET_SUBSCRIPTION_ENDED
+};
+
+struct aws_rr_subscription_status_event {
+    enum aws_rr_subscription_event_type type;
+    struct aws_byte_cursor topic_filter;
+    uint64_t operation_id;
+};
+
+/*
+ * Invariant: despite being on the same thread, these callbacks must be queued as cross-thread tasks on the native
+ * request-response client.  This allows us to iterate internal collections without worrying about external
+ * callers disrupting things by invoking APIs back on us.
+ */
+typedef void(
+    aws_rr_subscription_status_event_callback_fn)(const struct aws_rr_subscription_status_event *event, void *userdata);
+
+struct aws_rr_subscription_manager_options {
+
+    /*
+     * Maximum number of concurrent subscriptions allowed
+     */
+    size_t max_subscriptions;
+
+    /*
+     * Ack timeout to use for all subscribe and unsubscribe operations
+     */
+    uint32_t operation_timeout_seconds;
+
+    aws_rr_subscription_status_event_callback_fn *subscription_status_callback;
+    void *userdata;
+};
+
+/*
+ * The subscription manager works from a purely lazy perspective.  Unsubscribes (from topic filters that are no longer
+ * referenced) occur when looking for new subscription space.  Unsubscribe failures don't trigger anything special,
+ * we'll just try again next time we look for subscription space.  Subscribes are attempted on idle subscriptions
+ * that still need them, either in response to a new operation listener or a connection resumption event.
+ *
+ * We only allow one subscribe or unsubscribe to be outstanding at once for a given topic.  If an operation requires a
+ * subscription while an unsubscribe is in progress the operation is blocked until the unsubscribe resolves.
+ *
+ * These invariants are dropped during shutdown.  In that case, we immediately send unsubscribes for everything
+ * that is not already unsubscribing.
+ */
+struct aws_rr_subscription_manager {
+    struct aws_allocator *allocator;
+
+    struct aws_rr_subscription_manager_options config;
+
+    /* non-owning reference; the client is responsible for destroying this asynchronously (listener detachment) */
+    struct aws_mqtt_protocol_adapter *protocol_adapter;
+
+    /* &aws_rr_subscription_record.topic_filter_cursor -> aws_rr_subscription_record * */
```

Comment:
Needs cleanup?

Reply:
Not following; that's a type definition doc comment

Reply:
It's a type annotation.